source/model/backtraceMetrics.ts

```diff
@@ -0,0 +1,228 @@
+import { BacktraceClientOptions } from '..';
+import { currentTimestamp, getBacktraceGUID, getEndpointParams, uuid } from '../utils';
+import { APP_NAME, USER_AGENT, VERSION } from '../consts/application';
+import { SEC_TO_MILLIS } from '../consts';
+
+/**
+ * Handles Backtrace Metrics.
+ */
+export class BacktraceMetrics {
+  private readonly universe: string;
+  private readonly token: string;
+  private readonly timeout: number  = 15000; // Fifteen seconds in milliseconds.
+  private readonly persistenceInterval: number  = 1800000; // Thirty minutes in milliseconds.
+  private readonly heartbeatInterval: number  = 60000; // One minutes in milliseconds.
+
+  private readonly timestamp = currentTimestamp()
+  private readonly userAgent = USER_AGENT;
+  private readonly applicationName = APP_NAME;
+  private readonly applicationVersion = VERSION;
+
+  private readonly guid = getBacktraceGUID();
+
+  public hostname = 'https://events.backtrace.io';
+
+  constructor(configuration: BacktraceClientOptions) {
+    if (!configuration.endpoint) {
+      throw new Error(`Backtrace: missing 'endpoint' option.`);
+    }
+    const {universe, token} = getEndpointParams(configuration.endpoint)
+
+    if(!universe) {
+      throw new Error(`Backtrace: 'universe' could not be parsed from the endpoint.`);
+    }
+
+    if(!token && !configuration.token) {
+      throw new Error(`Backtrace: missing 'token' option or it could not be parsed from the endpoint.`);
+    }
+
+    this.universe = universe;
+    this.token = configuration.token || token || ''; // default '' case needed since type checking isn't certain that type can not be undefined.
+    this.timeout = configuration.timeout;  
+
+    this.persistSession(); // Create/persist session on construction.
+    // Persist session if page is focused on heartbeat interval
+    setInterval(() => this.persistIfFocused(), this.heartbeatInterval);
```

Comment:
one more idea - if we use setInterval, we also should store the interval id somewhere. Why? We can also subscribe to the window close event and stop the interval from there. Why? Because we might have a situation when someone will have two the same pages of the same app in the webbrowser and in this case we send data twice. At least we can try to unsubscribe when we close the page.

Reply:
I made it now keep track of the intervalId in local storage. If a new page is opened, it will clear the existing interval before starting the new one.